### PR TITLE
Thapi Minor refactor

### PR DESCRIPTION
**Pull request by TApplencourt**
This will help later when need to do the same to change the source with `archive.`
(not that this removes a limitation of before who was useless)

### CODE REVIEW

File: utils/babeltrace_thapi.in [hunk 1]
```diff
@@ -430,11 +430,10 @@ $thapi_graph = { 'tally' => ['source.ctf.fs', 'filter.utils.muxer', 'filter.inte
...
                                 'sink.btx_timeline.timeline'],
                  'trace' => ['source.ctf.fs', 'filter.utils.muxer', 'sink.text.rubypretty'],
-                 'trace_live' => ['source.ctf.lttng_live', 'filter.utils.muxer', 'sink.text.rubypretty'],
                  'to_interval' => ['source.ctf.fs', 'filter.utils.muxer', 'filter.intervals.interval',
                                    'filter.btx_stripper.stripper', 'sink.ctf.fs'],
                  'to_aggreg' => ['source.ctf.fs', 'filter.utils.muxer', 'filter.intervals.interval',
                                  'filter.btx_aggreg.aggreg', 'filter.btx_stripper.stripper', 'sink.ctf.fs'] }
 
-command = 'trace_live' if command == 'trace' && $options[:live]
+$thapi_graph[command][0] = 'source.ctf.lttng_live' if $options[:live]
```
[the remark this answers is not in the document]
Comment (TApplencourt) [reply utils/babeltrace_thapi.in:438]: It's the easiest. If not, it makes the case statement when looping of the graph painful. 
I can create 2 dicts if we want ( $thapi_graph_default, and then $thapi_graph) but I don't really saw the point or create a function call `get_graph` who does the mutation internally return this. and then assign the output to a global variable.

[the remark this answers is not in the document]
Comment (TApplencourt) [reply utils/babeltrace_thapi.in]: Used a function in [3a8b4b9](https://github.com/argonne-lcf/THAPI/pull/261/commits/3a8b4b9d0e3e69681ab4a36778e4e2a157e5ebe5)

[the remark this answers is not in the document]
Comment (Kerilk) [reply utils/babeltrace_thapi.in]: I was more thinking of:
```Ruby
$thapi_graph = { 'tally' => ['filter.utils.muxer', 'filter.intervals.interval',
                             'filter.btx_aggreg.aggreg', 'sink.btx_tally.tally'],
                 'timeline' => ['filter.utils.muxer', 'filter.intervals.interval',
                                'sink.btx_timeline.timeline'],
                 'trace' => ['filter.utils.muxer', 'sink.text.rubypretty'],
                 'to_interval' => ['filter.utils.muxer', 'filter.intervals.interval',
                                   'filter.btx_stripper.stripper', 'sink.ctf.fs'],
                 'to_aggreg' => ['filter.utils.muxer', 'filter.intervals.interval',
                                 'filter.btx_aggreg.aggreg', 'filter.btx_stripper.stripper', 'sink.ctf.fs'] }

source_comp = if $options[:live]
                'source.ctf.lttng_live'
              else
                'source.ctf.fs'
              end

bt_graph = [source_comp] + $thapi_graph[command]
create_and_run_graph(command, bt_graph, ARGV)
```

And add `bt_graph` as an argument to `no_gc`

[the remark this answers is not in the document]
Comment (TApplencourt) [reply utils/babeltrace_thapi.in]: Yes same idea. I guess modifying the `commands / graph`versus the `graphs` dict. Now that that everything is encapsulated I don't see a big difference.

Far better than beffer indeed.

[the remark this answers is not in the document]
Comment (TApplencourt) [reply utils/babeltrace_thapi.in]: I can pass only the list of commands/ bt_grapg, but then I still need to pass the `command` for logging purpose (`      name = "babeltrace_thapi_cli_#{command}.sh"`). It feel weird to me

[the remark this answers is not in the document]
Comment (Kerilk) [reply utils/babeltrace_thapi.in]: yes I know the duplications feels strange, but less so to me than recrafting the list all the time during the recursive calls, where the graph is immutable

[the remark this answers is not in the document]
Comment (TApplencourt) [reply utils/babeltrace_thapi.in]: Fair point.
- I can cacheit, as it's immutable now. 
- Or pass the graphs dict from the top `create_and_run_graph`

[the remark this answers is not in the document]
Comment (Kerilk) [reply utils/babeltrace_thapi.in]: > 
>     * Or can indeed pass the `command` + `graph`.
> 

Looks better for me, for no.

[the remark this answers is not in the document]
Comment (TApplencourt) [reply utils/babeltrace_thapi.in]: Done by 7041ec9

File: utils/babeltrace_thapi.in [hunk 2]
```diff
@@ -273,10 +273,28 @@ def modify_metadata(command, trace)
...
+  { 'tally' => ['filter.utils.muxer', 'filter.intervals.interval',
+                'filter.btx_aggreg.aggreg', 'sink.btx_tally.tally'],
+    'timeline' => ['filter.utils.muxer', 'filter.intervals.interval',
+                   'sink.btx_timeline.timeline'],
+    'trace' => ['filter.utils.muxer', 'sink.text.rubypretty'],
+    'to_interval' => ['filter.utils.muxer', 'filter.intervals.interval',
+                      'filter.btx_stripper.stripper', 'sink.ctf.fs'],
+    'to_aggreg' => ['filter.utils.muxer', 'filter.intervals.interval', 'filter.btx_aggreg.aggreg',
+                    'filter.btx_stripper.stripper', 'sink.ctf.fs'] }.transform_values { |l| [source_comp] + l }
+end
```
Comment (Kerilk) [inline utils/babeltrace_thapi.in:292]:
```suggestion
$thapi_graphs = { 'tally' => ['source.ctf.fs', 'filter.utils.muxer', 'filter.intervals.interval',
                             'filter.btx_aggreg.aggreg', 'sink.btx_tally.tally'],
                 'timeline' => ['source.ctf.fs', 'filter.utils.muxer', 'filter.intervals.interval',
                                'sink.btx_timeline.timeline'],
                 'trace' => ['source.ctf.fs', 'filter.utils.muxer', 'sink.text.rubypretty'],
                 'trace_live' => ['source.ctf.lttng_live', 'filter.utils.muxer', 'sink.text.rubypretty'],
                 'to_interval' => ['source.ctf.fs', 'filter.utils.muxer', 'filter.intervals.interval',
                                   'filter.btx_stripper.stripper', 'sink.ctf.fs'],
                 'to_aggreg' => ['source.ctf.fs', 'filter.utils.muxer', 'filter.intervals.interval',
                                 'filter.btx_aggreg.aggreg', 'filter.btx_stripper.stripper', 'sink.ctf.fs'] }

def bt_graph(command)
  source_comp = if $options[:live]
                  'source.ctf.lttng_live'
                else
                  'source.ctf.fs'
                end
  [source_comp] + $thapi_graphs[command]
end
```

What do you think of this?

File: utils/babeltrace_thapi.in [hunk 3]
```diff
@@ -424,17 +442,4 @@ $options[:'backend-names'] = $options[:backends].map { |name_level| name_level.s
...
-                 'trace' => ['source.ctf.fs', 'filter.utils.muxer', 'sink.text.rubypretty'],
-                 'trace_live' => ['source.ctf.lttng_live', 'filter.utils.muxer', 'sink.text.rubypretty'],
-                 'to_interval' => ['source.ctf.fs', 'filter.utils.muxer', 'filter.intervals.interval',
-                                   'filter.btx_stripper.stripper', 'sink.ctf.fs'],
-                 'to_aggreg' => ['source.ctf.fs', 'filter.utils.muxer', 'filter.intervals.interval',
-                                 'filter.btx_aggreg.aggreg', 'filter.btx_stripper.stripper', 'sink.ctf.fs'] }
-
-command = 'trace_live' if command == 'trace' && $options[:live]
-create_and_run_graph(command, ARGV)
+create_and_run_graph(command, bt_graphs[command], ARGV)
```
Comment (Kerilk) [inline utils/babeltrace_thapi.in:445]:
```suggestion
create_and_run_graph(command, bt_graph(command), ARGV)
```
Also needed if suggestion above is applied.

File: utils/babeltrace_thapi.in (hunk 2 above)
Comment (Kerilk) [reply utils/babeltrace_thapi.in:292]: I don't like the function instantiating a big data structure.

Comment (Kerilk) [reply utils/babeltrace_thapi.in:292]: or make `$thapi_graphs` a constant `THAPI_GRAPHS`.

Comment (TApplencourt) [reply utils/babeltrace_thapi.in:292]: I don't like have `THAPI_GRAPHS` as a variable in global scope. As it should't not be used, it need to be modified by `bt_graph` to be "complete"  and usable.

File: utils/babeltrace_thapi.in (hunk 3 above)
Comment (TApplencourt) [reply utils/babeltrace_thapi.in:447]: Graph cached, so built only once. Same memory footprint as a static global variable

**Review by Kerilk — APPROVED** [review]

**Pull request merged**